### PR TITLE
Changed the logging for config service to try to limit deadlock or live lock issues

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java [hunk 1]
```diff
@@ -67,23 +66,24 @@
...
+    @GuardedBy("this")
     private State state = State.INITIALIZED;
     private volatile EndpointMapping mapping = EndpointMapping.EMPTY;
 
     public enum SyncStatus { NOT_STARTED, NOTIFYING, COMPLETED }
 
     static class EpochState extends AbstractConfigurationService.AbstractEpochState
     {
-        SyncStatus syncStatus = SyncStatus.NOT_STARTED;
+        @GuardedBy("AccordConfigurationService.this")
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java]: this is checked then we update disk state... so sharing the lock keeps reasoning about correctness simpler

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java [hunk 2]
```diff
@@ -325,7 +325,7 @@ private void reportMetadata(ClusterMetadata metadata)
         Stage.MISC.submit(() -> reportMetadataInternal(metadata));
     }
 
-    synchronized void reportMetadataInternal(ClusterMetadata metadata)
+    void reportMetadataInternal(ClusterMetadata metadata)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:327]: the real `reportMetadataInternal` is already `synchronized`

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java [hunk 3]
```diff
@@ -360,7 +360,7 @@ private static boolean shareShard(Topology current, Node.Id target, Node.Id self
         return false;
     }
 
-    public synchronized void onNodesRemoved(long epoch, Topology current, Set<Node.Id> removed)
+    public void onNodesRemoved(long epoch, Topology current, Set<Node.Id> removed)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:369]: `syncPropagator` manages its own locks, so can ignore.

`nonCompletedEpochsBefore` and `receiveRemoteSyncCompletePreListenerNotify` also manage their own locks

File: test/distributed/org/apache/cassandra/distributed/impl/IsolatedExecutor.java [hunk 4]
```diff
@@ -282,7 +282,8 @@ public static <T> T waitOn(Future<T> f)
         }
         catch (ExecutionException e)
         {
-            throw Throwables.throwAsUncheckedException(e.getCause());
+//            throw Throwables.throwAsUncheckedException(e.getCause());
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/impl/IsolatedExecutor.java]: let me revert....

File: test/simulator/test/org/apache/cassandra/simulator/test/EpochStressTest.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+ --add-opens java.base/jdk.internal.module=ALL-UNNAMED
+ --add-opens java.base/jdk.internal.util.jar=ALL-UNNAMED
+ --add-opens jdk.management/com.sun.management.internal=ALL-UNNAMED
+ --add-opens jdk.management.jfr/jdk.management.jfr=ALL-UNNAMED
+ --add-opens java.desktop/com.sun.beans.introspect=ALL-UNNAMED
+ */
+public class EpochStressTest extends SimulationTestBase
+{
+    @Test
+    public void manyEpochsAndAccordConverges() throws IOException
```
Comment (dcapwell) [inline test/simulator/test/org/apache/cassandra/simulator/test/EpochStressTest.java:128]: reviewer: I have no idea what I am doing... so please let me know anything useful for working with simulated tests...

File: test/simulator/test/org/apache/cassandra/simulator/test/EpochStressTest.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                     return ActionList.of(actions);
+                 },
+                 config -> config.nodes(3, 3)
+                                 .dcs(1, 1)
+                                 .threadCount(100));
+    }
+
+    private static void validate()
+    {
+        Logger logger = LoggerFactory.getLogger(EpochStressTest.class);
```
Comment (dcapwell) [inline test/simulator/test/org/apache/cassandra/simulator/test/EpochStressTest.java:165]: I don't know why but having it in static seemed to loose the logs, but putting it here I got the logs... I have no clue why....

Comment (dcapwell) [reply test/simulator/test/org/apache/cassandra/simulator/test/EpochStressTest.java:165]: I do see several runs with no logs, this worries me...

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java (hunk 2 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:327]: shouldn't we be applying the same treatment to `reportMetadataInternal`? It looks to invoke several callbacks while holding the monitor.

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java [hunk 7]
```diff
@@ -381,7 +381,7 @@ public synchronized void onNodesRemoved(long epoch, Topology current, Set<Node.I
     private long[] nonCompletedEpochsBefore(long max)
```
Comment (belliottsmith) [inline src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:387]: I don't think it is safe for this method to be called from an unsynchronised block without itself being synchronised

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java (hunk 2 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:327]: this is called from 2 places: startup, TCM notification.  On TCM notification we are on `MiscStage` which is single threaded, so the sync isn't buying us anything there.

On startup, there are 2 different code paths: before TCM listen, after.

1) before we register with TCM the caller is the main thread, so no concurrency here. Sync doesn't buy us much
2) after register with TCM we check if there is a epoch gap between what we *think* the latest is and what TCM thinks.  If there is a gap *we* need to resolve it (else accord knows epochs [5, 6, 10])... In this case `reportTopology` is safe to call in any order, so is `onNodesRemoved`, so that really boils down to `updateMapping` which is already getting called in different orders *and* also syncs...


TL;DR - Yes, I think it's safe to drop here.  Ill try to think through this more to help confirm

Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:327]: I just noticed something I didn't see when I wrote the logic

```
Topology current = isEmpty() ? Topology.EMPTY : currentTopology();
```

in the case I listed above (`[5, 6, 10]`) the gaps we are filling would be for topologies in the past, so the `onNodesRemoved` doesn't make sense.  Also, *should* this step run if the epoch is further ahead?  so if we see 12 before `11` is received should we also call?  the locking doesn't really change this but think I need to flesh this part out more

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java (hunk 7 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:387]: there are 3 things that matter:

1) `epochs.minEpoch()`
2) `epochs.maxEpoch()`
3) `snapshot.syncStatus`

first 2 share the epoch lock, and `maxEpoch` could grow between calls to `minEpoch`, but for this context that seems fine.

That leaves `snapshot.syncStatus`, which does need the `AccordConfigurationService.this` lock for visibility...

To limit the locks as much as possible I could lock `getEpochSnapshot` or I can just make syncStatus volatile... given that this is the only code path that isn't accessed with a lock and its via `getEpochSnapshot` which is only called here, I went with sync on `getEpochSnapshot`.

A 3rd option is just accept that its stale... by the time this method exits the status might flip to complete, so we double add the node to `remote_sync_complete` in the topologies table, since this is a set it will dedup for us.

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java (hunk 2 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:327]: removed sync and fixed the race condition bug with onRemove checking... we now wait for the epoch to be ack, then we compare against `epoch - 1`, this will enforce only 1 epoch will trigger the remove

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java [hunk 8]
```diff
@@ -67,23 +66,24 @@
...
     private volatile EndpointMapping mapping = EndpointMapping.EMPTY;
 
     public enum SyncStatus { NOT_STARTED, NOTIFYING, COMPLETED }
 
     static class EpochState extends AbstractConfigurationService.AbstractEpochState
     {
-        SyncStatus syncStatus = SyncStatus.NOT_STARTED;
+        @GuardedBy("AccordConfigurationService.this")
+        private SyncStatus syncStatus = SyncStatus.NOT_STARTED;
         protected final AsyncResult.Settable<Void> localSyncNotified = AsyncResults.settable();
```
Comment (belliottsmith) [inline src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:86]: is it safe that we set this (and notify its callbacks) while holding the `AccordConfigurationService` lock?

Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:86]: It looks like today this is only used in test methods, but I would prefer that we rename methods that can register callbacks as unsafe, and document why they are potentially unsafe, as it could cause surprises later.

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java [hunk 9]
```diff
@@ -381,7 +388,7 @@ public synchronized void onNodesRemoved(long epoch, Topology current, Set<Node.I
     private long[] nonCompletedEpochsBefore(long max)
     {
         LongArrayList notComplete = new LongArrayList();
-        for (long epoch = epochs.minEpoch(); epoch <= max && epoch <= epochs.maxEpoch(); epoch++)
+        for (long epoch = epochs.minEpoch(), maxKnown = epochs.maxEpoch(); epoch <= max && epoch <= maxKnown; epoch++)
```
Comment (belliottsmith) [inline src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java]: I don't think this iteration is safe? `minEpoch` and `maxEpoch` can change while the loop is running. I think we need the epoch monitor when we run this loop.

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java [hunk 10]
```diff
@@ -614,28 +636,32 @@ public static EpochSnapshot notStarted(long epoch)
...
+    public EpochSnapshot getEpochSnapshot(long epoch)
     {
         if (epoch < epochs.minEpoch() || epoch > epochs.maxEpoch())
             return null;
 
-        return new EpochSnapshot(getOrCreateEpochState(epoch));
+        // lock for read visibility to syncStatus
+        synchronized (this)
+        {
+            return new EpochSnapshot(getOrCreateEpochState(epoch));
```
Comment (belliottsmith) [inline src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java]: do we want to hold both locks at once? can't we get the `EpochState` and then build `EpochSnapshot`? What visibility are we looking for here, isn't it enough to piggy-back on the visibility semantics of the `epoch` monitor?

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java (hunk 8 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:86]: renamed the public api to `unsafeLocalSyncNotified` and added a javadoc saying why it is unsafe

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java (hunk 9 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java]: I do think we might due to the fact `accord.impl.AbstractConfigurationService.AbstractEpochHistory#truncateUntil` could be called and `accord.impl.AbstractConfigurationService.AbstractEpochHistory#getOrCreate` can re-populate the past epochs.

so added the lock

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java (hunk 10 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java]: called `getOrCreateEpochState` outside the lock

> What visibility are we looking for her

Most call paths of `org.apache.cassandra.service.accord.AccordConfigurationService.EpochState#syncStatus` already hold the accord lock, so need this here else it can be stale (unless we make it volatile)

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java (hunk 7 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java:388]: We can prune, so min epoch can move also.

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java (hunk 10 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java]: Technically it has visibility due to the semantics of the `getOrCreateEpochState` - monitor entry has the semantics of a global read barrier. But this isn't important to optimise, and it's fine to make volatile.

File: src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java [hunk 11]
```diff
@@ -660,6 +660,11 @@ public long maxEpoch()
         return epochs.maxEpoch();
     }
 
+    /**
+     * The callback is resolved while holding the class lock, which can cause the future chain to resolve while also
```
Comment (belliottsmith) [inline src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java]: object lock, not class lock?

**Review by belliottsmith — APPROVED** [review]
